File: lib/models/rvc.dart
/// Registro de Visita a Cliente (RVC)
class Rvc {
  final String id;
  final String vendedorId;
  final DateTime fecha;
  final String zona;
  final int clientesVisitados;
  final int clientes60Visitados;
  final int clientesPerdidosVisitados;
  final double ventaTotal;
  final double recaudoTotal;
  final List<String> clientesNoVisitados;
  final bool descuentosAplicados;

  Rvc({
    required this.id,
    required this.vendedorId,
    required this.fecha,
    required this.zona,
    this.clientesVisitados = 0,
    this.clientes60Visitados = 0,
    this.clientesPerdidosVisitados = 0,
    this.ventaTotal = 0,
    this.recaudoTotal = 0,
    this.clientesNoVisitados = const [],
    this.descuentosAplicados = false,
  });

  Map<String, dynamic> toMap() => {
        'id': id,
        'vendedorId': vendedorId,
        'fecha': fecha.toIso8601String().split('T')[0],
        'zona': zona,
        'clientesVisitados': clientesVisitados,
        'clientes60Visitados': clientes60Visitados,
        'clientesPerdidosVisitados': clientesPerdidosVisitados,
        'ventaTotal': ventaTotal,
        'recaudoTotal': recaudoTotal,
        'clientesNoVisitados': clientesNoVisitados.join(','),
        'descuentosAplicados': descuentosAplicados ? 1 : 0,
      };

  factory Rvc.fromMap(Map<String, dynamic> map) => Rvc(
        id: map['id'] as String,
        vendedorId: map['vendedorId'] as String,
        fecha: DateTime.parse(map['fecha'] as String),
        zona: map['zona'] as String? ?? '',
        clientesVisitados: map['clientesVisitados'] as int? ?? 0,
        clientes60Visitados: map['clientes60Visitados'] as int? ?? 0,
        clientesPerdidosVisitados:
            map['clientesPerdidosVisitados'] as int? ?? 0,
        ventaTotal: (map['ventaTotal'] ?? 0).toDouble(),
        recaudoTotal: (map['recaudoTotal'] ?? 0).toDouble(),
        clientesNoVisitados: (map['clientesNoVisitados'] as String?)
                ?.split(',')
                .where((s) => s.isNotEmpty)
                .toList() ??
            [],
        descuentosAplicados: (map['descuentosAplicados'] ?? 0) == 1,
      );
}

File: api/endpoints/rvc.php
<?php
declare(strict_types=1);

function handleRvc($conn, string $method, ?string $id): void
{
    if ($method === 'GET') {
        $fecha = $_GET['fecha'] ?? date('Y-m-d');
        $vendedorId = isset($_GET['vendedorId']) ? trim((string)$_GET['vendedorId']) : '';

        if ($vendedorId !== '') {
            $row = dbQueryOne(
                $conn,
                'SELECT * FROM rvc WHERE vendedorId = ? AND fecha = ?',
                [$vendedorId, $fecha]
            );
            jsonResponse($row ?: null);
            return;
        }

        $rows = dbQueryAll($conn, 'SELECT * FROM rvc WHERE fecha = ?', [$fecha]);
        jsonResponse($rows);
        return;
    }

    if ($method === 'POST') {
        $d = getJsonInput();
        $rid = optionalString($d, 'id', '');
        if ($rid === '') {
            $rid = uniqid('rvc_', false);
        }

        $payload = [
            requiredString($d, 'vendedorId'),
            requiredString($d, 'fecha'),
            optionalString($d, 'zona', ''),
            (int)($d['clientesVisitados'] ?? 0),
            (int)($d['clientes60Visitados'] ?? 0),
            (int)($d['clientesPerdidosVisitados'] ?? 0),
            (float)($d['ventaTotal'] ?? 0),
            (float)($d['recaudoTotal'] ?? 0),
            optionalString($d, 'clientesNoVisitados', ''),
            (int)($d['descuentosAplicados'] ?? 0),
        ];

        dbBegin($conn);
        try {
            $exists = dbQueryOne($conn, 'SELECT COUNT(1) AS total FROM rvc WHERE id = ?', [$rid]);
            if (((int)($exists['total'] ?? 0)) > 0) {
                dbExecute(
                    $conn,
                    'UPDATE rvc SET vendedorId=?,fecha=?,zona=?,clientesVisitados=?,clientes60Visitados=?,clientesPerdidosVisitados=?,ventaTotal=?,recaudoTotal=?,clientesNoVisitados=?,descuentosAplicados=? WHERE id=?',
                    array_merge($payload, [$rid])
                );
                dbCommit($conn);
                jsonResponse(['success' => true, 'id' => $rid, 'mode' => 'updated']);
                return;
            }

            dbExecute(
                $conn,
                'INSERT INTO rvc (id,vendedorId,fecha,zona,clientesVisitados,clientes60Visitados,clientesPerdidosVisitados,ventaTotal,recaudoTotal,clientesNoVisitados,descuentosAplicados) VALUES (?,?,?,?,?,?,?,?,?,?,?)',
                array_merge([$rid], $payload)
            );
            dbCommit($conn);
            jsonResponse(['success' => true, 'id' => $rid, 'mode' => 'inserted']);
            return;
        } catch (Throwable $e) {
            dbRollback($conn);
            throw $e;
        }
    }

    jsonResponse(['success' => false, 'error' => 'Método no permitido'], 405);
}
